### PR TITLE
[pt1][quant] QTensor permute, transpose and contiguous

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20869 [pt1][quant] QTensor permute, transpose and contiguous**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15474060/)

Adding support for the functions listed in the title, by implementing the copy kernel.

Differential Revision: [D15474060](https://our.internmc.facebook.com/intern/diff/D15474060/)

### CODE REVIEW

File: aten/src/ATen/native/cpu/CopyKernel.cpp [hunk 1]
```diff
@@ -29,6 +29,12 @@ static void copy_kernel(TensorIterator& iter, bool non_blocking) {
   if (dtype == iter.dtype(1)) {
     if (dtype == ScalarType::Half) {
       unary_kernel(iter, [=](at::Half a) -> at::Half { return a; });
+    } else if (isQIntType(dtype)) {
```
Comment (dzhulgakov) [inline aten/src/ATen/native/cpu/CopyKernel.cpp]: is the reason you can't use AT_DISPATCH_ALL_TYPES_AND2 below that vectorized things are not defined for qtypes? might be worth to define them - but if it's too hard, I'm fine doing it in a separate diff

File: aten/src/ATen/native/Copy.cpp [hunk 2]
```diff
@@ -107,8 +107,15 @@ Tensor & copy_(Tensor & self, const Tensor & src, bool non_blocking) {
...
   }
 
-  if (self.scalar_type() == kQUInt8) {
-    return quantized_copy_(self, src);
+  if (self.is_quantized() && !src.is_quantized()) {
+    return quantized_copy_from_float_(self, src);
+  }
+
+  if (self.is_quantized() && src.is_quantized()) {
+    TORCH_CHECK(self.q_scale().toFloat() == src.q_scale().toFloat(),
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/Copy.cpp]: yeah, will add a TODO

File: aten/src/ATen/native/TensorProperties.cpp [hunk 3]
```diff
@@ -60,7 +60,13 @@ Tensor contiguous(const Tensor& self, MemoryFormat memory_format) {
   if (self.is_contiguous(memory_format)) {
     return self;
   }
-  auto result = at::empty_like(self);
+  Tensor result;
+  if (self.is_quantized()) {
+    // FIXME: only works for PerTensorAffineQuantized now
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/TensorProperties.cpp]: we don't have qscheme() yet, will add a TODO

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 4]
```diff
@@ -49,10 +49,11 @@ Quantizer* quantizer(const Tensor& self) {
 
 Tensor int_repr_quant(const Tensor& self) {
   Tensor dst;
+  auto self_c = self.contiguous();
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/QTensor.cpp]: sure. does tensoriterator work with this type of copy?

File: aten/src/ATen/native/cpu/CopyKernel.cpp (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/native/cpu/CopyKernel.cpp]: Yes. sure I can try that in another diff

File: aten/src/ATen/native/TensorProperties.cpp [hunk 5]
```diff
@@ -60,7 +60,13 @@ Tensor contiguous(const Tensor& self, MemoryFormat memory_format) {
...
+    result = at::_empty_affine_quantized(self.sizes(), self.options(), self.q_scale().toDouble(), self.q_zero_point().toLong());
+  } else {
+    result = at::empty_like(self);
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/TensorProperties.cpp]: for `empty` we have a split, but `empty_like` should be fine I think, will merge into `empty_like`

**Review by dzhulgakov — APPROVED** [review]
Looks good, but there are some build issues

**Pull request closed without merging**